prefect_aws/ecs.py

```diff
@@ -0,0 +1,981 @@
+"""
+Examples:
+
+    Run a task using ECS Fargate
+    >>> ECSTask(command=["echo", "hello world"]).run()
+
+    Run a task using ECS Fargate with a spot container instance
+    >>> ECSTask(command=["echo", "hello world"], launch_type="FARGATE_SPOT").run()
+
+    Run a task using ECS with an EC2 container instance
+    >>> ECSTask(command=["echo", "hello world"], launch_type="EC2").run()
+
+    Run a task on a specific VPC using ECS Fargate
+    >>> ECSTask(command=["echo", "hello world"], vpc_id="vpc-01abcdf123456789a").run()
+
+    Run a task and stream the container's output to the local terminal. Note an
+    execution role must be provided with permissions: logs:CreateLogStream,
+    logs:CreateLogGroup, and logs:PutLogEvents.
+    >>> ECSTask(
+    >>>     command=["echo", "hello world"],
+    >>>     stream_output=True,
+    >>>     execution_role_arn="..."
+    >>> )
+
+    Run a task using an existing task definition as a base
+    >>> ECSTask(command=["echo", "hello world"], task_definition_arn="arn:aws:ecs:...")
+
+    Run a task with a specific image
+    >>> ECSTask(command=["echo", "hello world"], image="alpine:latest")
+
+    Run a task with custom memory and CPU requirements
+    >>> ECSTask(command=["echo", "hello world"], memory=4096, cpu=2048)
+
+    Run a task with custom environment variables
+    >>> ECSTask(command=["echo", "hello $PLANET"], env={"PLANET": "earth"})
+
+    Run a task in a specific ECS cluster
+    >>> ECSTask(command=["echo", "hello world"], cluster="my-cluster-name")
+"""
+import copy
+import sys
+import time
+import warnings
+from typing import Any, Dict, Generator, List, Optional, Tuple
+
+import boto3
+import yaml
+from anyio.abc import TaskStatus
+from prefect.docker import get_prefect_image_name
+from prefect.infrastructure.base import Infrastructure, InfrastructureResult
+from prefect.utilities.asyncutils import run_sync_in_worker_thread, sync_compatible
+from pydantic import Field, root_validator
+from typing_extensions import Literal
+
+from prefect_aws import AwsCredentials
+
+# Internal type alias for ECS clients which are generated dynamically in botocore
+_ECSClient = Any
+
+
+class ECSTaskResult(InfrastructureResult):
+    """The result of a run of an ECS task"""
```

Comment:
I'm leaving room to attach infrastructure specific details to them without changing the return signature. Since user's aren't using these blocks directly yet, we haven't gotten any requests, but I'm sure we will. In hindsight, it might not have been worth introducing the subtypes yet but it's still nice and clear.